### ISSUE TITLE
Unusual behavior using manual instrumentation of distributed traces using OpenTracer in v1.x

### ISSUE THREAD

**Issue by okhwaja** [DataDog/dd-trace-rb#2191]
Hi,

We’re encountering an issue when trying to doing the major version upgrade of our codebase from ddtrace v0.54.2 to v1.2.0

We’re using OpenTracing and doing manual instrumentation between different types of services: apis —> jobs. On the jobs side, it doesn’t seem like we’re able to successfully resume a trace like we were in the 0.54.x version.

**Current behaviour**
When attempting to upgrade to v1.2.0, we are missing many spans from our downstream job execution (activesupport, memcached, etc.) and two of our custom traces are out of order
<img width="701" alt="image" src="https://user-images.githubusercontent.com/3255907/181815163-919997a7-f83a-4833-a5d2-53baf706ad58.png">

**Expected behaviour**
On the v0.54.2 version of ddtrace, we see the expected trace hierarchy and spans from many integrations
<img width="1429" alt="image" src="https://user-images.githubusercontent.com/3255907/181814881-828023a9-a5a5-4542-9f9d-a9d161491d62.png">

**Steps to reproduce**

I’ve written a small spec that fails that might be related to the problem - it seems like there’s an issue using the extracted trace context from `OpenTracing.extract(…)`. https://github.com/okhwaja/ddtrace-bug is the repo and the test code is here: https://github.com/okhwaja/ddtrace-bug/blob/main/spec/v1_repro_spec.rb. The setup instructions are in the README

One odd thing I noticed when looking in code, [the RackPropagator always returns a nil value for datadog_context](https://cs.github.com/DataDog/dd-trace-rb/blob/f1a8956baa5c1f552c89f8a6c72cc04e075e5d5a/lib/datadog/opentracer/rack_propagator.rb#L53-L57) since it seems a migration was made to use trace_digests instead. Then in `Datadog::OpenTracer::Tracer.start_active_span`, [we use that always nil value](https://cs.github.com/DataDog/dd-trace-rb/blob/f1a8956baa5c1f552c89f8a6c72cc04e075e5d5a/lib/datadog/opentracer/tracer.rb#L95-L99). Note that if `child_of` were not passed in, [the code would generate a span with a not-null datadog_context](https://cs.github.com/DataDog/dd-trace-rb/blob/f1a8956baa5c1f552c89f8a6c72cc04e075e5d5a/lib/datadog/opentracer/tracer.rb#L81) and use that instead.

Happy to provide more information if you need it! Appreciate any guidance or tips you’re able to provide - thanks!

**How does `ddtrace` help you?**

**Environment**

* **ddtrace version:** v1.2.0
* **Configuration block (`Datadog.configure ...`):** I can provide a full version if needed - summarized version below
```
# Important: OpenTracing.global_tracer must be assigned before the Datadog.configure block.
datadog_opentracer = Datadog::OpenTracer::Tracer.new(
  enabled: true, ...
)
OpenTracing.global_tracer = datadog_opentracer

Datadog.configure do |c|
  c.tracing.instrument :rack, request_queuing: true,
    service_name: "#{@service_name}-rack",
    web_service_name: "#{@service_name}-proxy"
  c.tracing.instrument :rails, request_queuing: true

  c.tracing.instrument :active_support, cache_service: "#{@service_name}-dbcache"

  c.tracing.instrument :active_record,
    analytics_enabled: false,
    service_name: "#{@service_name}-activerecord"
  c.tracing.instrument :dalli,
    analytics_enabled: false
  c.tracing.instrument :elasticsearch,
    analytics_enabled: false
  c.tracing.instrument :faraday,
    analytics_enabled: false
  c.tracing.instrument :http,
    analytics_enabled: false,
    split_by_domain: _to_bool(ENV["DATADOG_APM_HTTP_SPLIT_BY_DOMAIN"])
  c.tracing.instrument :redis,
    analytics_enabled: false
  c.tracing.instrument :sidekiq,
    analytics_enabled: false,
    service_name: "#{@service_name}"
  c.tracing.instrument :aws
end
```
* **Ruby version:** v2.7.4
* **Operating system:**
* **Relevant library versions:** [see repro repo](https://github.com/okhwaja/ddtrace-bug/blob/main/example.gemspec#L34-L38)

**Comment by TonyCTHsu** [comment]
@okhwaja Thanks for the reporting. I will have a look!

**Comment by TonyCTHsu** [comment]
https://github.com/DataDog/dd-trace-rb/pull/2201#issuecomment-1204594941

👋 @okhwaja, I have taken a look into your concern about `start_active_span`, this should be fine. Since the empty `datadog_context` would be populated in [tracing](https://github.com/DataDog/dd-trace-rb/blob/master/lib/datadog/opentracer/tracer.rb#L146).

From the disjointing flamegraph you've shown, I think it likely is caused by the broken carrier propagation. 

The fix for propagation is on the way for `1.3.0`. Please do let me know if you are still encountering issue 😄

**Comment by okhwaja** [comment]
thanks @TonyCTHsu ! will try it out when 1.3.0 is released 🙇‍♂️

**Comment by TonyCTHsu** [comment]
@okhwaja `1.3.0` has released! Give it a try 🎉

**Comment by okhwaja** [comment]
hey @TonyCTHsu , sorry just picking this back up. Unfortunately I don't think this resolved our issue 🤔  still seeing orphaned spans and missing integrations like AWS, redis, etc. 

(after)
<img width="826" alt="image" src="https://user-images.githubusercontent.com/3255907/187266464-2060f9a8-af45-40eb-a693-f8b8d8c8b92f.png">

(before)

<img width="838" alt="image" src="https://user-images.githubusercontent.com/3255907/187266649-e549e2c4-da12-47d0-af70-63b817b96c69.png">

This is a manual instrumentation. Ultimately we're calling
```
parent_span_context = OpenTracing.extract(OpenTracing::FORMAT_RACK, headers)
scope = OpenTracing.start_active_span(span_name, child_of: parent_span_context, start_time: start_time, tags: tags)
```

I shelled into the host to see an example `parent_span_context`
```
> parent_span_context
=> #<Datadog::OpenTracer::SpanContext:0x0000557ed79c8b90 @datadog_context=nil, @datadog_trace_digest=#<Datadog::Tracing::TraceDigest:0x0000557ed79c8fa0 @span_id=3055899650488270179, @span_name=nil, @span_resource=nil, @span_service=nil, @span_type=nil, @trace_hostname=nil, @trace_id=2418612052018279426, @trace_name=nil, @trace_origin=nil, @trace_process_id=nil, @trace_resource=nil, @trace_runtime_id=nil, @trace_sampling_priority=1, @trace_service=nil>, @baggage={}>
```

should we be using some other method to manually instrument? I'm still sniffing around [this section of code](https://github.com/DataDog/dd-trace-rb/pull/2201#issuecomment-1204594941) since the `datadog_context` is null.

**Comment by TonyCTHsu** [comment]
@okhwaja Thanks for reporting, I am looking into this!

**Comment by TonyCTHsu** [comment]
I was able to reproduce the such kinda a flamegraph pattern with https://github.com/DataDog/dd-trace-rb/compare/master...tonycthsu/opentracing-distributed-tracing

![Screenshot 2022-09-01 at 13 28 43](https://user-images.githubusercontent.com/16049123/188097088-1fb4cdca-fef1-4f15-a73f-e9e9a91748b4.png)

The bug seems to be related to referring to stale span context during manual instrumentation with Opentracing, I have failing test ready and currently working on the fix on [this](https://github.com/DataDog/dd-trace-rb/pull/2253)

**Comment by marcotc** [comment]
👋 @okhwaja, we implemented a fix for distributed traces for OpenTracer in this PR: #2289

Could you try our prerelease gem, which includes that fix, and let us know if the problem is addressed for you?
```ruby
source 'http://gems.datadoghq.com/prerelease-v2' do
  gem 'ddtrace', '1.4.1.fix.opentracing.context.propagation.271910'
end
```

Let us know if you have any further questions!

**Comment by okhwaja** [comment]
thanks @marcotc and @delner ! Clearly word got to you faster than I did but I tested this morning and I'm seeing all the expected spans and usual hierarchy

<img width="861" alt="image" src="https://user-images.githubusercontent.com/3255907/192391102-dce481da-389a-40e5-90c4-dca6a954a959.png">

Will test out the final version when a new release gets cut!

**Comment by marcotc** [comment]
@okhwaja this change is now released in `v1.4.2`, give it a try!
